### ISSUE TITLE
[BUG] Swiping through images or opening large threads in quick succession breaks image loading with media prefetch enabled

### ISSUE THREAD

**Issue by Astridchan** [Adamantcheese/Kuroba#150]
Steps to reproduce (with media prefetch enabled):
Bug 1:
1. Open thread with many images (some thread on /wg/ for example)
2. After threads open, immediately start swiping through images without waiting for them to load
3. Some images you swipe through will fail to load, the loading progress bar will get stuck at around 10%. Those images will not load even if you refresh the thread or close and reopen Kuroba.

Bug in action:
https://streamable.com/hbd1q

Bug 2:
1. Open thread with many images (some thread on /wg/ for example)
2. After thread opens, go through a few images to confirm images are loading, then go back to catalog and open another thread
3. After opening 1-3 threads with several images posted in them, images will fail to load in any successive threads you open.

Bug in action:
https://streamable.com/4ttxe

Possibly related to #114.

**Comment by Adamantcheese** [comment]
Likely related to #56 and the associated "images fail to load sometimes", which was even before prefetching was implemented. I'm thinking it's an issue with the file cache implementation at this point.

**Comment by ownBard** [comment]
Thanks, I think this is more accurate of the main issues I've been having with prefetch. I just clear the image cache in dev options whenever I have problems and that usually fixes it.

**Comment by Adamantcheese** [comment]
I noticed that you don't have wifi on, can you confirm this happens on wifi as well? And if you don't mind, could you confirm this happens on other sites as well? It may be Cloudflare doing things because there's a hell of a lot of packets being sent out with prefetch on.

**Comment by Astridchan** [comment]
Being on WiFi didn't make any difference. 

I've been playing around opening different threads and seeing how the images load, and I'm almost certain now that this bug is to do with how the prefetch process prioritises loading images. What made me suspect it in the first place was because I noticed that if you waited long enough, images that would fail to load initially would eventually get loaded. Basically, what is happening is when you open a thread is that the prefetch process starts loading the images in order from first to last:
Image 1
Image 2
Image 3
Image 4
...
Image 99

If you swipe past pictures before they have been loaded that makes the prefetch process skip them and keep going down the list loading images, but once it's got to the last image in the thread, it will then jump back to load the images that were skipped earlier:

Image 1 LOADED
Image 2 SKIPPED
Image 3 SKIPPED
Image 4 LOADED
Image 5 LOADED
...
Image 99 LOADED

Image 2 LOADED
Image 3 LOADED

Images that were skipped won't load until the prefetch process has finished loading the rest of the images in the thread. tested this multiple times, and every single time, once the last image in the thread got loaded, the images earlier in the thread that had failed to load would get loaded.

Same thing is happening with opening multiple threads and having images fail to load. Images in the second thread you open will not start loading until the prefetch process is finished loading the images in the first thread. I even noticed that if you open two threads in quick succession and some images fail to load in the first thread, the prefetch process will continue to the second thread before it loads the skipped images in the first thread, like this:

**Thread 1**
Image 1 LOADED
Image 2 SKIPPED
Image 3 SKIPPED
Image 4 LOADED
Image 5 LOADED
...
Image 99 LOADED

**Thread 2**
Image 1 LOADED
Image 2 LOADED
...
Image 99 LOADED

**Thread 1**
Image 2 LOADED
Image 3 LOADED

I think at least these conditions are required to be added to the media prefetching to fix this issue:
1. Always prioritize loading the image that is currently open
2. Always prioritize loading images in the thread that is currently opened

**Comment by Adamantcheese** [comment]
Alright, plenty of good info there. Couple of things though:
1) When you exit out of the pager view, all preloading images from the pager's point of view are cancelled (the one to the right and the current image). This means that if they were already being prefetched by the thread level prefetch, those will be cancelled as well because the file cache only keeps one downloader for two requests of the same asset. They'll be attempted to be prefetched again when the thread gets updated, or if you start swiping through images again. So your first suggestion is "already in place" and due to the nature of thread pools I have no control over the priority of when it gets executed in the thread pool. If prefetching is done for the thread, it'll be at the front of the queue and have priority. Otherwise it's wherever.

2) See end of above answer. Unless you want to write up a custom thread pool with priority controls for the file cache, that's just how it's going to be. When you load up thread 1, everything's added to the pool's queue, then thread 2's is added to the queue. It would probably be wise here to remove all downloaders from the queue that were for thread 1, but I don't really have any way of doing that without passing around huge amounts of useless data. And then moving to a priority pool, if I don't remove thread 1's downloaders, then we're at the same problem where they're at the same priority and it doesn't matter what's downloaded, from the pool's perspective.

Basically, the best I'm willing to do is change the thread pool size from a fixed limit of 2 to infinite as an attempted workaround. A priority thread pool would be the proper solution here but I don't use the feature and don't care much to actually implement more than this rudimentary version of it.

**Comment by ownBard** [comment]
If only it could prefetch in reverse order. It seems backwards to queue requests when you want the most recently selected images/threads to load first.

**Comment by Adamantcheese** [comment]
It depends on the type of thread you're viewing really. If it's a new thread and you want to see it all, first images first. If it's an old thread and you want to get to the end, last images first. There's no good way to make that distinction between the two, so an ideal solution is a two pronged approach with a priority pool, where image priority for a thread starts off very high at the first and last images of the thread and decreases going towards the center.

**Comment by ownBard** [comment]
I didn't mean bottom to top I meant last-in first-out — sorry if my English isn't great. But yes it seems it would need priority thread pool either way.

**Comment by Astridchan** [comment]
1. What I mean with "always prioritize loading the image that is currently open", is that if you open an image that hasn't already been loaded yet, the app should stop prefetching images in the background to load the currently opened image, and then resume loading the rest of the images. With how the prefetching currently works, you have to wait until all other images in the open thread (and all images in a second thread, if I you've opened a second thread) have loaded before the image you're trying to open will load. If you're browsing threads with a lot of images you'll be waiting for a long time. This really doesn't make much sense UX-wise. To use a music player analogy, the current situation is like if you were playing music from a playlist, and the only way you could jump to a song you wanted to play is to wait for all the other songs in the active playlist to play through first, whereas any proper music player will allow you to jump to play any song in the playlist, or even songs outside of the playlist, at any time.

2. >It would probably be wise here to remove all downloaders from the queue that were for thread 1, but I don't really have any way of doing that without passing around huge amounts of useless data.

I don't really  understand what you mean with "passing around huge amounts of useless data", but I really think what you're describing is exactly what should be happening; the prefetch process should only be queuing and loading images from a single thread at a time. So if I open thread 1, images from that thread start loading. If I then close the thread and open thread 2, the prefetch process will stop/pause loading images from thread 1 and start loading images from thread 2. If I close thread 2 and open thread 1 again, loading images from thread 2 will stop and loading them from thread 1 will resume.

**Comment by ownBard** [comment]
Well if all cache downloaders can be cancelled under those circumstances then yes, otherwise no.

**Comment by Adamantcheese** [comment]
A note for anyone if they want to implement a priority pool, the priority numbers for each image in a thread should follow this formula: `image_dl_priority = floor(priority_scale(cos(2pi*post_number_in_thread/post_count)/2 + .5) + 1)`; note that `priority_scale` should likely be `post_count` or greater, so that you get good separation of priority levels.

**Comment by Adamantcheese** [comment]
@Astridchan I don't think you understand at all how the prefetching currently works. Please go take a look at onChanLoaderData in ThreadPresenter. I didn't implement finer grained control than literally dumping everything into the file cache queue. A priority pool is the proper solution here. For 2. you're assuming that I have control over wildly different threads when I have no central mechanism to control what onChanLoaderData functions are currrently executing. I would need to pass around a lot more data to know what I should be loading at any given time.

**Comment by Pentaphon** [comment]
@Astridchan I have this problem where webms take forever but I don't have prefetch enabled.

**Comment by ownBard** [comment]
@Pentaphon That is a seperate issue (#56)

**Comment by Adamantcheese** [comment]
Closed with merge.
